// kysely/dist/cjs/query-executor/default-query-executor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultQueryExecutor = void 0;
const query_executor_js_1 = require("./query-executor.js");
class DefaultQueryExecutor extends query_executor_js_1.QueryExecutor {
    #compiler;
    #connectionProvider;
    constructor(compiler, connectionProvider, plugins = []) {
        super(plugins);
        this.#compiler = compiler;
        this.#connectionProvider = connectionProvider;
    }
    compileQuery(node) {
        return this.#compiler.compileQuery(node);
    }
    async executeQueryImpl(compiledQuery) {
        return this.#connectionProvider.withConnection((connection) => {
            return connection.executeQuery(compiledQuery);
        });
    }
    withPlugin(plugin) {
        return new DefaultQueryExecutor(this.#compiler, this.#connectionProvider, [
            ...this.plugins,
            plugin,
        ]);
    }
    withPluginAtFront(plugin) {
        return new DefaultQueryExecutor(this.#compiler, this.#connectionProvider, [
            plugin,
            ...this.plugins,
        ]);
    }
    withConnectionProvider(connectionProvider) {
        return new DefaultQueryExecutor(this.#compiler, connectionProvider, [
            ...this.plugins,
        ]);
    }
    withoutPlugins() {
        return new DefaultQueryExecutor(this.#compiler, this.#connectionProvider, []);
    }
}
exports.DefaultQueryExecutor = DefaultQueryExecutor;
