// kysely/dist/cjs/query-executor/noop-query-executor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopQueryExecutor = void 0;
const query_executor_js_1 = require("./query-executor.js");
/**
 * A {@link QueryExecutor} subclass that can be used when you don't
 * have a {@link QueryCompiler}, {@link ConnectionProvider} or any
 * other needed things to actually execute queries.
 */
class NoopQueryExecutor extends query_executor_js_1.QueryExecutor {
    compileQuery() {
        throw new Error('this query cannot be compiled to SQL');
    }
    async executeQueryImpl() {
        throw new Error('this query cannot be executed');
    }
    withConnectionProvider() {
        throw new Error('this query cannot have a connection provider');
    }
    withPlugin(plugin) {
        return new NoopQueryExecutor([...this.plugins, plugin]);
    }
    withPluginAtFront(plugin) {
        return new NoopQueryExecutor([plugin, ...this.plugins]);
    }
    withoutPlugins() {
        return new NoopQueryExecutor([]);
    }
}
exports.NoopQueryExecutor = NoopQueryExecutor;
