// kysely/dist/cjs/query-builder/join-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JoinBuilder = void 0;
const join_node_js_1 = require("../operation-node/join-node.js");
const filter_parser_js_1 = require("../parser/filter-parser.js");
const object_utils_js_1 = require("../util/object-utils.js");
const prevent_await_js_1 = require("../util/prevent-await.js");
class JoinBuilder {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    on(...args) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: join_node_js_1.JoinNode.cloneWithOn(this.#props.joinNode, (0, filter_parser_js_1.parseOnFilter)(this.#props.parseContext, args)),
        });
    }
    orOn(...args) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: join_node_js_1.JoinNode.cloneWithOrOn(this.#props.joinNode, (0, filter_parser_js_1.parseOnFilter)(this.#props.parseContext, args)),
        });
    }
    /**
     * Just like {@link WhereInterface.whereRef} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.whereRef} for documentation and examples.
     */
    onRef(lhs, op, rhs) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: join_node_js_1.JoinNode.cloneWithOn(this.#props.joinNode, (0, filter_parser_js_1.parseReferenceFilter)(this.#props.parseContext, lhs, op, rhs)),
        });
    }
    /**
     * Just like {@link WhereInterface.orWhereRef} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.orWhereRef} for documentation and examples.
     */
    orOnRef(lhs, op, rhs) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: join_node_js_1.JoinNode.cloneWithOrOn(this.#props.joinNode, (0, filter_parser_js_1.parseReferenceFilter)(this.#props.parseContext, lhs, op, rhs)),
        });
    }
    /**
     * Just like {@link WhereInterface.whereExists} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.whereExists} for documentation and examples.
     */
    onExists(arg) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: join_node_js_1.JoinNode.cloneWithOn(this.#props.joinNode, (0, filter_parser_js_1.parseExistFilter)(this.#props.parseContext, arg)),
        });
    }
    /**
     * Just like {@link WhereInterface.whereNotExists} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.whereNotExists} for documentation and examples.
     */
    onNotExists(arg) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: join_node_js_1.JoinNode.cloneWithOn(this.#props.joinNode, (0, filter_parser_js_1.parseNotExistFilter)(this.#props.parseContext, arg)),
        });
    }
    /**
     * Just like {@link WhereInterface.orWhereExists} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.orWhereExists} for documentation and examples.
     */
    orOnExists(arg) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: join_node_js_1.JoinNode.cloneWithOrOn(this.#props.joinNode, (0, filter_parser_js_1.parseExistFilter)(this.#props.parseContext, arg)),
        });
    }
    /**
     * Just like {@link WhereInterface.orWhereNotExists} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.orWhereNotExists} for documentation and examples.
     */
    orOnNotExists(arg) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: join_node_js_1.JoinNode.cloneWithOrOn(this.#props.joinNode, (0, filter_parser_js_1.parseNotExistFilter)(this.#props.parseContext, arg)),
        });
    }
    toOperationNode() {
        return this.#props.joinNode;
    }
}
exports.JoinBuilder = JoinBuilder;
(0, prevent_await_js_1.preventAwait)(JoinBuilder, "don't await JoinBuilder instances. They are never executed directly and are always just a part of a query.");
