// kysely/dist/cjs/dialect/mysql/mysql-introspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MysqlIntrospector = void 0;
const migrator_js_1 = require("../../migration/migrator.js");
const object_utils_js_1 = require("../../util/object-utils.js");
class MysqlIntrospector {
    #db;
    constructor(db) {
        this.#db = db;
    }
    async getMetadata(options = { withInternalKyselyTables: false }) {
        let query = this.#db
            .selectFrom('information_schema.columns')
            .selectAll()
            .where('table_schema', '=', this.#db.raw('database()'))
            .castTo();
        if (!options.withInternalKyselyTables) {
            query = query
                .where('table_name', '!=', migrator_js_1.MIGRATION_TABLE)
                .where('table_name', '!=', migrator_js_1.MIGRATION_LOCK_TABLE);
        }
        const rawColumns = await query.execute();
        return {
            tables: this.#parseTableMetadata(rawColumns),
        };
    }
    #parseTableMetadata(columns) {
        return columns.reduce((tables, it) => {
            let table = tables.find((tbl) => tbl.name === it.TABLE_NAME);
            if (!table) {
                table = (0, object_utils_js_1.freeze)({
                    name: it.TABLE_NAME,
                    columns: [],
                });
                tables.push(table);
            }
            table.columns.push((0, object_utils_js_1.freeze)({
                name: it.COLUMN_NAME,
                dataType: it.DATA_TYPE,
                isNullable: it.IS_NULLABLE === 'YES',
            }));
            return tables;
        }, []);
    }
}
exports.MysqlIntrospector = MysqlIntrospector;
