// @serverless-stack/resources/dist/DebugApp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugApp = void 0;
const cdk = __importStar(require("aws-cdk-lib"));
const Construct_1 = require("./Construct");
/**
 * The DebugApp construct is used internally by SST to
 * - Deploy the [`DebugStack`](DebugStack.md). It contains the resources that powers [Live Lambda Development](/live-lambda-development.md).
 * - Automatically prefix the debug stack name with the stage and app name.
 *
 * It extends [`cdk.App`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.App.html). It's made available as the `app` in the `debugApp()` callback in the `stacks/index.js` of your SST app.
 *
 * ```js
 * export function debugApp(app) {
 *   new sst.DebugStack(app, "debug-stack");
 * }
 * ```
 *
 * Since it is initialized internally, the props that are passed to `DebugApp` cannot be changed.
 *
 * @example
 */
class DebugApp extends cdk.App {
    /**
     * @internal
     */
    constructor(deployProps) {
        super();
        this.name = deployProps.name;
        this.stage = deployProps.stage;
        this.region = deployProps.region;
        this.account = process.env.CDK_DEFAULT_ACCOUNT || "my-account";
    }
    synth(options = {}) {
        // Check app has stack
        const stacks = this.node.children.filter((child) => (0, Construct_1.isSSTDebugStack)(child));
        if (stacks.length > 1) {
            console.error(`Error: You can only create 1 DebugStack inside the "debugApp()" callback.\n`);
            process.exit(1);
        }
        if (stacks.length === 0) {
            console.error(`Error: The "debugApp()" callback is not creating a DebugStack.\n`);
            process.exit(1);
        }
        for (const child of this.node.children) {
            if ((0, Construct_1.isStackConstruct)(child)) {
                // Stack names need to be parameterized with the stage name
                if (!child.stackName.startsWith(`${this.stage}-`) &&
                    !child.stackName.endsWith(`-${this.stage}`) &&
                    child.stackName.indexOf(`-${this.stage}-`) === -1) {
                    console.error(`Error: Stack "${child.stackName}" is not parameterized with the stage name. The stack name needs to either start with "$stage-", end in "-$stage", or contain the stage name "-$stage-".\n`);
                    process.exit(1);
                }
            }
        }
        return super.synth(options);
    }
    /**
     * Use this method to prefix resource names in your stacks to make sure they don't thrash when deployed to different stages in the same AWS account. This method will prefix a given resource name with the stage and app name. Using the format `${stage}-${name}-${logicalName}`.
     */
    logicalPrefixedName(logicalName) {
        return `${this.stage}-${this.name}-${logicalName}`;
    }
}
exports.DebugApp = DebugApp;
