// @serverless-stack/resources/dist/FunctionalStack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStack = exports.dependsOn = exports.use = exports.stack = void 0;
const Stack_1 = require("./Stack");
function stack(app, fn, props) {
    currentApp = app;
    currentStack = fn;
    const id = (props === null || props === void 0 ? void 0 : props.id) || fn.name;
    const exists = getExports(app).has(fn);
    if (exists)
        throw new Error(`StackDuplicates: Attempting to initialize stack ${id} several times`);
    const stack = new EmptyStack(app, id, props);
    getStacks(app).set(fn, stack);
    const ctx = {
        app,
        stack,
    };
    const returns = fn.bind(stack)(ctx);
    if (returns && "then" in returns)
        return returns.then((data) => {
            getExports(app).set(fn, data);
        });
    getExports(app).set(fn, returns);
    return app;
}
exports.stack = stack;
let currentApp;
let currentStack;
const exportsCache = new Map();
const stackCache = new Map();
function getExports(app) {
    if (!exportsCache.has(app))
        exportsCache.set(app, new Map());
    return exportsCache.get(app);
}
function getStacks(app) {
    if (!stackCache.has(app))
        stackCache.set(app, new Map());
    return stackCache.get(app);
}
function use(stack) {
    if (!currentApp)
        throw new Error("No app is set");
    const exports = getExports(currentApp);
    if (!exports.has(stack))
        throw new Error(`StackWrongOrder: Initialize "${stack.name}" stack before "${currentStack === null || currentStack === void 0 ? void 0 : currentStack.name}" stack`);
    return exports.get(stack);
}
exports.use = use;
function dependsOn(stack) {
    const current = getStack(currentStack);
    const target = getStack(stack);
    current.addDependency(target);
}
exports.dependsOn = dependsOn;
function getStack(stack) {
    if (!currentApp)
        throw new Error("No app is set");
    const stacks = getStacks(currentApp);
    if (!stacks.has(stack))
        throw new Error(`StackWrongOrder: Initialize "${stack.name}" stack before "${currentStack === null || currentStack === void 0 ? void 0 : currentStack.name}" stack`);
    return stacks.get(stack);
}
exports.getStack = getStack;
class EmptyStack extends Stack_1.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
    }
}
