// @serverless-stack/resources/dist/Topic.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Topic = void 0;
const constructs_1 = require("constructs");
const sns = __importStar(require("aws-cdk-lib/aws-sns"));
const snsSubscriptions = __importStar(require("aws-cdk-lib/aws-sns-subscriptions"));
const Construct_1 = require("./Construct");
const Function_1 = require("./Function");
const Queue_1 = require("./Queue");
/////////////////////
// Construct
/////////////////////
/**
 * The `Topic` construct is a higher level CDK construct that makes it easy to create a serverless pub/sub service. You can create a topic that has a list of subscribers. And you can publish messages to it from any part of your serverless app.
 *
 * You can have two types of subscribers; Function subscribers (subscribe with a Lambda function) or Queue subscribers (subscribe with a SQS queue).
 *
 * This construct makes it easier to define a topic and its subscribers. It also internally connects the subscribers and topic together.
 *
 * @example
 * ### Using the minimal config
 *
 * ```js
 * import { Topic } from "@serverless-stack/resources";
 *
 * new Topic(stack, "Topic", {
 *   subscribers: {
 *     subscriber1: "src/function1.handler",
 *     subscriber2: "src/function2.handler"
 *   },
 * });
 * ```
 */
class Topic extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        this.props = props || {};
        this.cdk = {};
        this.subscribers = {};
        this.permissionsAttachedForAllSubscribers = [];
        this.createTopic();
        this.addSubscribers(this, (props === null || props === void 0 ? void 0 : props.subscribers) || {});
    }
    /**
     * The ARN of the internally created SNS Topic.
     */
    get topicArn() {
        return this.cdk.topic.topicArn;
    }
    /**
     * The name of the internally created SNS Topic.
     */
    get topicName() {
        return this.cdk.topic.topicName;
    }
    /**
     * Get a list of subscriptions for this topic
     */
    get subscriptions() {
        return Object.values(this.subscribers).map((sub) => {
            let children;
            // look for sns.Subscription inside Queue.sqsQueue
            if (sub instanceof Queue_1.Queue) {
                children = sub.cdk.queue.node.children;
            }
            // look for sns.Subscription inside Function
            else {
                children = sub.node.children;
            }
            const child = children.find((child) => {
                return (0, Construct_1.isCDKConstructOf)(child, "aws-cdk-lib.aws_sns.Subscription");
            });
            return child;
        });
    }
    /**
     * A list of the internally created function instances for the subscribers.
     */
    get subscriberFunctions() {
        return Object.values(this.subscribers).filter((subscriber) => subscriber instanceof Function_1.Function);
    }
    /**
     * Add subscribers to the topic.
     *
     * @example
     * ```js {5}
     * const topic = new Topic(stack, "Topic", {
     *   subscribers: {
     *     subscriber1: "src/function1.handler",
     *     subscriber2: "src/function2.handler"
     *   },
     * });
     * topic.addSubscribers(stack, {
     *   subscriber3: "src/function3.handler"
     * });
     * ```
     */
    addSubscribers(scope, subscribers) {
        Object.entries(subscribers).forEach(([subscriberName, subscriber]) => {
            this.addSubscriber(scope, subscriberName, subscriber);
        });
    }
    /**
     * Attaches the given list of permissions to all the subscriber functions. This allows the subscribers to access other AWS resources.
     *
     * @example
     *
     * ```js
     * const topic = new Topic(stack, "Topic", {
     *   subscribers: {
     *     subscriber1: "src/function1.handler",
     *     subscriber2: "src/function2.handler"
     *   },
     * });
     * topic.attachPermissions(["s3"]);
     * ```
     */
    attachPermissions(permissions) {
        Object.values(this.subscribers)
            .filter((subscriber) => subscriber instanceof Function_1.Function)
            .forEach((subscriber) => subscriber.attachPermissions(permissions));
        this.permissionsAttachedForAllSubscribers.push(permissions);
    }
    /**
     * Attaches the list of permissions to a given subscriber by index
     * @example
     * ```js {5}
     * const topic = new Topic(stack, "Topic", {
     *   subscribers: {
     *     subscriber1: "src/function1.handler",
     *     subscriber2: "src/function2.handler"
     *   },
     * });
     *
     * topic.attachPermissionsToSubscriber("subscriber1", ["s3"]);
     * ```
     */
    attachPermissionsToSubscriber(subscriberName, permissions) {
        const subscriber = this.subscribers[subscriberName];
        if (!(subscriber instanceof Function_1.Function)) {
            throw new Error(`Cannot attach permissions to the "${this.node.id}" Topic subscriber because it's not a Lambda function`);
        }
        subscriber.attachPermissions(permissions);
    }
    getConstructMetadata() {
        return {
            type: "Topic",
            data: {
                topicArn: this.cdk.topic.topicArn,
                // TODO: Deprecate eventually and mirror KinesisStream
                subscribers: Object.values(this.subscribers).map(Construct_1.getFunctionRef),
                subscriberNames: Object.keys(this.subscribers),
            },
        };
    }
    createTopic() {
        const app = this.node.root;
        const { cdk } = this.props;
        if ((0, Construct_1.isCDKConstruct)(cdk === null || cdk === void 0 ? void 0 : cdk.topic)) {
            this.cdk.topic = cdk === null || cdk === void 0 ? void 0 : cdk.topic;
        }
        else {
            const snsTopicProps = ((cdk === null || cdk === void 0 ? void 0 : cdk.topic) || {});
            this.cdk.topic = new sns.Topic(this, "Topic", Object.assign({ topicName: app.logicalPrefixedName(this.node.id) }, snsTopicProps));
        }
    }
    addSubscriber(scope, subscriberName, subscriber) {
        if (subscriber instanceof Queue_1.Queue ||
            subscriber.queue) {
            subscriber = subscriber;
            this.addQueueSubscriber(scope, subscriberName, subscriber);
        }
        else {
            subscriber = subscriber;
            this.addFunctionSubscriber(scope, subscriberName, subscriber);
        }
    }
    addQueueSubscriber(scope, subscriberName, subscriber) {
        var _a;
        // Parse subscriber props
        let subscriptionProps;
        let queue;
        if (subscriber instanceof Queue_1.Queue) {
            subscriber = subscriber;
            queue = subscriber;
        }
        else {
            subscriber = subscriber;
            subscriptionProps = (_a = subscriber.cdk) === null || _a === void 0 ? void 0 : _a.subscription;
            queue = subscriber.queue;
        }
        this.subscribers[subscriberName] = queue;
        // Create Subscription
        this.cdk.topic.addSubscription(new snsSubscriptions.SqsSubscription(queue.cdk.queue, subscriptionProps));
    }
    addFunctionSubscriber(scope, subscriberName, subscriber) {
        var _a, _b;
        // Parse subscriber props
        let subscriptionProps;
        let functionDefinition;
        if (typeof subscriber !== "string" && "function" in subscriber) {
            subscriptionProps = (_a = subscriber.cdk) === null || _a === void 0 ? void 0 : _a.subscription;
            functionDefinition = subscriber.function;
        }
        else {
            subscriber = subscriber;
            functionDefinition = subscriber;
        }
        // Create function
        const fn = Function_1.Function.fromDefinition(scope, `Subscriber_${this.node.id}_${subscriberName}`, functionDefinition, (_b = this.props.defaults) === null || _b === void 0 ? void 0 : _b.function, `The "defaults.function" cannot be applied if an instance of a Function construct is passed in. Make sure to define all the subscribers using FunctionProps, so the Topic construct can apply the "defaults.function" to them.`);
        this.subscribers[subscriberName] = fn;
        // Create Subscription
        this.cdk.topic.addSubscription(new snsSubscriptions.LambdaSubscription(fn, subscriptionProps));
        // Attach existing permissions
        this.permissionsAttachedForAllSubscribers.forEach((permissions) => fn.attachPermissions(permissions));
    }
}
exports.Topic = Topic;
