// @serverless-stack/resources/dist/ViteStaticSite.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViteStaticSite = void 0;
const path = __importStar(require("path"));
const fs = __importStar(require("fs-extra"));
const StaticSite_1 = require("./StaticSite");
/////////////////////
// Construct
/////////////////////
/**
 * The `ViteStaticSite` construct is a higher level CDK construct that makes it easy to create a Vite single page app. It provides a simple way to build and deploy the site to an S3 bucket; setup a CloudFront CDN for fast content delivery; and configure a custom domain for the website URL.
 *
 * It's designed to work with static sites built with [Vite](https://vitejs.dev/). It allows you to [automatically set environment variables](#configuring-environment-variables) in your Vite app directly from the outputs of your SST app. And it can also create a `.d.ts` type definition file for the environment variables.
 *
 * The `ViteStaticSite` construct internally extends the [`StaticSite`](StaticSite.md) construct with the following pre-configured defaults.
 *
 * - [`indexPage`](StaticSite.md#indexpage) is set to `index.html`.
 * - [`errorPage`](StaticSite.md#errorpage) is set to `redirect_to_index_page`. So error pages are redirected to the index page.
 * - [`buildCommand`](StaticSite.md#buildcommand) is `npm run build`.
 * - [`buildOutput`](StaticSite.md#buildoutput) is the `dist` folder in your Vite app.
 * - [`fileOptions`](StaticSite.md#fileoptions) sets the cache control to `max-age=0,no-cache,no-store,must-revalidate` for HTML files; and `max-age=31536000,public,immutable` for JS/CSS files.
 *
 * @example
 *
 * ### Minimal Config
 *
 * Deploys a Vite app in the `path/to/src` directory.
 *
 * ```js
 * new ViteStaticSite(stack, "Site", {
 *   path: "path/to/src",
 * });
 * ```
 */
class ViteStaticSite extends StaticSite_1.StaticSite {
    constructor(scope, id, props) {
        const { path: sitePath, environment, typesPath } = props || {};
        // generate buildCommand
        let defaultBuildCommand = "npm run build";
        if (fs.existsSync(path.join(sitePath, "yarn.lock"))) {
            defaultBuildCommand = "yarn build";
        }
        // create types file
        const filePath = path.resolve(path.join(sitePath, typesPath || "src/sst-env.d.ts"));
        generateTypesFile(filePath, environment);
        super(scope, id, Object.assign({ indexPage: "index.html", errorPage: "redirect_to_index_page", buildCommand: defaultBuildCommand, buildOutput: "dist", fileOptions: [
                {
                    exclude: "*",
                    include: "*.html",
                    cacheControl: "max-age=0,no-cache,no-store,must-revalidate",
                },
                {
                    exclude: "*",
                    include: ["*.js", "*.css"],
                    cacheControl: "max-age=31536000,public,immutable",
                },
            ] }, props));
    }
}
exports.ViteStaticSite = ViteStaticSite;
function generateTypesFile(typesFullPath, environment) {
    const content = `/// <reference types="vite/client" />

interface ImportMetaEnv {
${Object.keys(environment || {})
        .map((key) => `  readonly ${key}: string`)
        .join("\n")}
}

interface ImportMeta {
  readonly env: ImportMetaEnv
}`;
    fs.ensureDirSync(path.dirname(typesFullPath));
    fs.writeFileSync(typesFullPath, content);
}
