// kysely/dist/cjs/operation-node/create-table-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateTableNode = void 0;
const object_utils_js_1 = require("../util/object-utils.js");
/**
 * @internal
 */
exports.CreateTableNode = (0, object_utils_js_1.freeze)({
    is(node) {
        return node.kind === 'CreateTableNode';
    },
    create(table) {
        return (0, object_utils_js_1.freeze)({
            kind: 'CreateTableNode',
            table,
            columns: (0, object_utils_js_1.freeze)([]),
        });
    },
    cloneWithColumn(createTable, column) {
        return (0, object_utils_js_1.freeze)({
            ...createTable,
            columns: (0, object_utils_js_1.freeze)([...createTable.columns, column]),
        });
    },
    cloneWithConstraint(createTable, constraint) {
        return (0, object_utils_js_1.freeze)({
            ...createTable,
            constraints: createTable.constraints
                ? (0, object_utils_js_1.freeze)([...createTable.constraints, constraint])
                : (0, object_utils_js_1.freeze)([constraint]),
        });
    },
    cloneWith(createTable, params) {
        return (0, object_utils_js_1.freeze)({
            ...createTable,
            ...params,
        });
    },
});
