// @serverless-stack/core/dist/state/state.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.suggestStage = exports.setStage = exports.getStage = exports.directory = exports.init = exports.stacksPath = exports.resolve = exports.Function = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
const aws_sdk_1 = require("aws-sdk");
exports.Function = __importStar(require("./function"));
function resolve(root, relative) {
    return path_1.default.resolve(root, ".sst", relative);
}
exports.resolve = resolve;
function stacksPath(root) {
    return path_1.default.join(root, ".build", "lib");
}
exports.stacksPath = stacksPath;
function init(root) {
    const dir = directory(root);
    if (fs_1.default.existsSync(dir))
        return;
    fs_1.default.mkdirSync(dir, { recursive: true });
}
exports.init = init;
function directory(root) {
    return resolve(root, "");
}
exports.directory = directory;
function resolveStage(root) {
    return resolve(root, "stage");
}
function getStage(root) {
    const file = resolveStage(root);
    try {
        return fs_1.default.readFileSync(file).toString().trim();
    }
    catch {
        return null;
    }
}
exports.getStage = getStage;
function setStage(root, stage) {
    const file = resolveStage(root);
    fs_1.default.writeFileSync(file, stage);
}
exports.setStage = setStage;
async function suggestStage() {
    const client = new aws_sdk_1.STS();
    const result = await client
        .getCallerIdentity()
        .promise()
        .then((result) => path_1.default.basename(result.Arn))
        .catch(() => os_1.default.userInfo().username);
    return result.replace(/[^A-Za-z0-9]/g, "-");
}
exports.suggestStage = suggestStage;
