// @serverless-stack/core/dist/telemetry/environment.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnvironmentData = void 0;
const os_1 = __importDefault(require("os"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const ci_info_1 = __importDefault(require("ci-info"));
let data;
function getEnvironmentData() {
    if (data) {
        return data;
    }
    const cpus = os_1.default.cpus() || [];
    const isSeed = !!process.env.SEED_APP_NAME;
    data = {
        // Software information
        systemPlatform: os_1.default.platform(),
        systemRelease: os_1.default.release(),
        systemArchitecture: os_1.default.arch(),
        // Machine information
        cpuCount: cpus.length,
        cpuModel: cpus.length ? cpus[0].model : null,
        cpuSpeed: cpus.length ? cpus[0].speed : null,
        memoryInMb: Math.trunc(os_1.default.totalmem() / Math.pow(1024, 2)),
        // Environment information
        isCI: ci_info_1.default.isCI || isSeed,
        ciName: (isSeed ? "Seed" : ci_info_1.default.name) || null,
        sstVersion: getSSTVersion(),
    };
    return data;
}
exports.getEnvironmentData = getEnvironmentData;
function getSSTVersion() {
    const module = "@serverless-stack/core/package.json";
    const packageJson = fs_extra_1.default.readJsonSync(require.resolve(module));
    return packageJson.version;
}
