// @serverless-stack/core/dist/packager/packager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getManager = void 0;
const cross_spawn_1 = __importDefault(require("cross-spawn"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
function Implementation(impl) {
    return impl;
}
const NPM = Implementation({
    type: "npm",
    add(opts) {
        return cross_spawn_1.default.sync("npm", [
            "install",
            "--save-exact",
            ...opts.pkgs,
            opts.type === "dependencies" ? "--save" : "--save-dev",
        ], {
            cwd: opts.cwd,
            stdio: opts.verbose ? "inherit" : undefined,
        });
    },
});
const Yarn = Implementation({
    type: "yarn",
    add(opts) {
        return cross_spawn_1.default.sync("yarn", [
            "add",
            "--exact",
            (opts.type === "devDependencies" && "--dev") || "",
            "-W",
            ...opts.pkgs,
        ].filter((item) => item), {
            cwd: opts.cwd,
            stdio: opts.verbose ? "inherit" : undefined,
        });
    },
});
function getManager(dir) {
    const lock = path_1.default.join(dir, "yarn.lock");
    if (fs_1.default.existsSync(lock))
        return Yarn;
    const upDir = path_1.default.resolve(dir, "..");
    if (upDir === dir) {
        return NPM;
    }
    return getManager(upDir);
}
exports.getManager = getManager;
